src/http/data_loader.rs

```diff
@@ -42,57 +49,95 @@ impl<C: HttpClient + Send + Sync + 'static + Clone> Loader<DataLoaderRequest> fo
     keys: &[DataLoaderRequest],
   ) -> async_graphql::Result<HashMap<DataLoaderRequest, Self::Value>, Self::Error> {
     if let Some(group_by) = self.batched.clone() {
-      let mut keys = keys.to_vec();
-      keys.sort_by(|a, b| a.to_request().url().cmp(b.to_request().url()));
+      // store the preused length of our keys.
+      let original_key_length = keys.len();
 
       let mut request = keys[0].to_request();
+
       let first_url = request.url_mut();
 
-      for key in &keys[1..] {
-        let request = key.to_request();
-        let url = request.url();
+      let urls = keys[1..].iter().map(|key: &DataLoaderRequest| {
+        let key_request_result = key.to_request();
+        tokio::task::spawn(async move {
+          let url = key_request_result.url().to_owned();
+          url
+        })
+      });
+
+      for url_handle in urls {
+        //TODO: replace this unsafe unwrap, and handle the error
+        let url = url_handle.await.unwrap();
         first_url.query_pairs_mut().extend_pairs(url.query_pairs());
       }
 
       let res = self.client.execute(request).await?;
+
       #[allow(clippy::mutable_key_type)]
-      let mut hashmap: HashMap<DataLoaderRequest, Response> = HashMap::with_capacity(keys.len());
-      let path = &group_by.path();
-      let body_value = res.body.group_by(path);
+      let mut hashmap: HashMap<DataLoaderRequest, Response> = HashMap::with_capacity(original_key_length);
+
+      //TODO: figure out something lifetime elision raised here due to 'life0 mismatch
+      let group_key = group_by.clone().key().to_string();
+      let path = group_by.path();
+      let body_value = res.body.group_by(&path[..]);
+
+      let mut res_handle = Vec::with_capacity(original_key_length);
 
-      for key in &keys {
+      for key in keys.iter() {
         let req = key.to_request();
-        let query_set: std::collections::HashMap<_, _> = req.url().query_pairs().collect();
-        let id = query_set
-          .get(group_by.key())
-          .ok_or(anyhow::anyhow!("Unable to find key {} in query params", group_by.key()))?;
+        let key_cloned = key.clone();
+        let group_key_clone = group_key.clone();
```

Comment:
yeah you are technically not incorrect this did crossed my mind multiple times while modifying the logic except for `group_key` it is the way that the internal schematics of crate are designed so can't do anything much there and since this loader method returns a std::collections::HashsMap<K, V> here it makes things even more challenging though, but I got another intresting plan for it here though ! I will give it a shot , these changes as I mentioned are currently very experimental from my side.

Reply:
I think you're pretty limited when working with async blocks and need to pass ownership anyway, so your best option to avoid cloning would be to move origin object

Reply:
@meskill I have revised those changes in 4d34aab3a4a1b4368d4cb3113c8437380967e734 to avoid any unnecessary cloning of data.